### PR TITLE
ENH: add supxlabel and supylabel

### LINKED ISSUE #11147: FR: add a supxlabel and supylabel as the suptitle function which are already exist
I did not manage to find if supxlabel/supylabel exist, but I think it should be a direct extension to suptitle (for subplots) and might be worth implementing.

Thanks.

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

UPDATE: 10 Oct 2020, this still works as below.

There are a couple of fixes to padding in `suptitle` placement in `constrained_layout`.  

This PR adds `supxlabel` and `supylabel`.    Works with tight_layout and the new constrained_layout.  Not so well with no layout manager ;-)

Closes #11147

```python
import matplotlib.pyplot as plt

for tl, cl in zip([True, False, False], [False, False, True]):
    fig = plt.figure(constrained_layout=cl, tight_layout=tl)

    gs = fig.add_gridspec(2, 3)

    ax = dict()

    ax['A'] = fig.add_subplot(gs[0, 0:2])
    ax['B'] = fig.add_subplot(gs[1, 0:2])
    ax['C'] = fig.add_subplot(gs[:, 2])

    ax['C'].set_xlabel('Booger')
    ax['B'].set_xlabel('Booger')
    ax['A'].set_ylabel('Booger Y')
    fig.suptitle(f'TEST: tight_layout={tl} constrained_layout={cl}')
    fig.supxlabel('XLAgg')
    fig.supylabel('YLAgg')

    fig.savefig(f'/Users/jklymak/downloads/suplabels_tl{tl}_cl{cl}.png', dpi=75)
    plt.show()

```
suplabels_tlFalse_clFalse
suplabels_tlTrue_clFalse
suplabels_tlFalse_clTrue

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by timhoffm** [comment]
Have you considered that people might want suplabels on parts of a grid? e.g. columns 1 and 2 share an xlabel while column 0 has it's own label.

I'm not claiming that you should implement it now or even that it's necessary, but the desire may come up. If we consider this a reasonable extension, we should at least think about if an extension of the proposed API in that direction is possible and staight forward.

**Comment by jklymak** [comment]
Right now the API is the same as `suptitle`, and that has the same potential to only apply to some axes and not others.  

There are two somewhat orthogonal issues here.  
  1.  suplabels don’t align with axes, they align w/ the figure.  It’s a reasonable extension to align them with the spines of the child axes, but I’ve not implemented this here.  It’s non trivial to do so in a way that works with all the layout managers...
  2. How do you only label a subset of axes?   I’d argue this is best served with the subpanels/subfigure concept.  Then the subpanel gets the suplabels.  I think we should make subpanels subclasses of figures (in some way) so they will get this change for free.

**Comment by jklymak** [comment]
i.e. this figure just has suptitles for each subpanel, but was made w two subpanels, and could just as easily have supx and supylabels.  

https://user-images.githubusercontent.com/1562854/82160762-74a30a80-984c-11ea-9cf1-025515b4d5b9.png

**Comment by sks-2020** [comment]
Could the functions to add supxlabels and supylabels please be added to future releases of gridspec? Also have multicolumn row titles in the gridspec layout, as in the below figure? 

image

**Comment by jklymak** [comment]
@sks-2020 What do you anticipate wanting to do with gridspec that the sub-panel functionality won't give you?

**Comment by sks-2020** [comment]
@jklymak I think the subpanel function will solve what I am hoping to do. My issue is the same as outlined here:
https://stackoverflow.com/questions/40936729/matplotlib-title-spanning-two-or-any-number-of-subplot-columns

I found a possible way of solving this, but doesn't use the neat layout option of using gridspec:
https://stackoverflow.com/questions/27426668/row-titles-for-matplotlib-subplot?rq=1

In summary, I have a gridspec subplots layout within another gridspec. I do the plotting part in the inner grid and need the labeling flexibility in the outer grid. At the moment, using the current release of matplotlib, I can only achieve this using text annotations. Inbuilt functions (like supxlabel, supylabel, supxticklabels, supyticklabels, suptitle), for the outer grid, will be very useful.

**Comment by jklymak** [comment]
@mwaskom This is the suplabel PR.  This doesn't *really* require the other PR, but would be nice to do it after....

**Comment by mwaskom** [comment]
I'm afraid I don't know where to find it in the diff, but is the default font size here bigger than regular axes labels?

I'm not sure I've ever needed to add a "suplabel" pertaining to multiple axes that have existing, distinct labels. My use-case is almost always a situation where I have multiple subplots with shared axes and I want to add a single label to identify them. In that case, I'd want to the label to have the same size as a normal axis label.

Also, will these suplabels be tighter against the axes when there are no first-order labels?

I won't assert or claim to know which is a more common usecase, but I do think it's an application this PR should consider...

**Comment by jklymak** [comment]
I think they may be larger by default (just like suptitle) but the user can change this. 

If there are no individual labels and you use constrained_layout the suplabels will be tight against the axes. Can likely be made to work with tight_layout as well.  With no layout manager the whole thing is a bit of a mess anyhow.  

As noted above, the suplables will be centered on the figure, not the spines or some other measure of what the label is supposed to span.  That may or may not be acceptable depending on axes decorations, but it's a lot simpler.

**Comment by jklymak** [comment]
This is ready for review, and closes at least one feature request, and something seaborn would like.

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -861,3 +861,48 @@ def test_subfigure_ss():
...
 
     fig.suptitle('Figure suptitle', fontsize='xx-large')
+
+
+@image_comparison(['test_subfigure_double.png'], style='mpl20',
+                  savefig_kwarg={'facecolor': 'teal'},
+                  remove_text=False)
+def test_subfigure_double():
+    # test assigning the subfigure via subplotspec
+    np.random.seed(19680808)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:871]: Not sure why the figure tests use the wrong date.
```suggestion
    np.random.seed(19680801)
```

File: lib/matplotlib/tests/test_figure.py [hunk 2]
```diff
@@ -861,3 +861,48 @@ def test_subfigure_ss():
...
+
+    subfigs[1].set_facecolor('coral')
+    subfigs[1].suptitle('subfigs[1]')
+
+    subfigsnest = subfigs[0].subfigures(2, 1, height_ratios=[1, 1.4])
+    subfigsnest[0].suptitle('subfigsnest[0]')
+    subfigsnest[0].set_facecolor('r')
+    axsnest0 = subfigsnest[0].subplots(1, 2, sharey=True)
+    for nn, ax in enumerate(axsnest0):
+        pc = example_plot(ax, hide_labels=False)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:897]: `nn` is unused:
```suggestion
    for ax in axsnest0:
        pc = example_plot(ax, hide_labels=False)
```

File: lib/matplotlib/tests/test_figure.py [hunk 3]
```diff
@@ -861,3 +861,48 @@ def test_subfigure_ss():
...
     fig.suptitle('Figure suptitle', fontsize='xx-large')
+
+
+@image_comparison(['test_subfigure_double.png'], style='mpl20',
+                  savefig_kwarg={'facecolor': 'teal'},
+                  remove_text=False)
+def test_subfigure_double():
+    # test assigning the subfigure via subplotspec
+    np.random.seed(19680808)
+    def example_plot(ax, fontsize=12, hide_labels=False):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:872]: `hide_labels` is never `True`, and this is only called once; maybe it should be embedded, and then `nn` could be used in the loop?

File: lib/matplotlib/tight_layout.py [hunk 4]
```diff
@@ -108,20 +108,31 @@ def auto_adjust_subplotpars(
...
                         + pad_inches / fig_width_inch)
     if not margin_top:
         margin_top = (max(vspaces[0, :].max(), 0)
                       + pad_inches / fig_height_inch)
-        suptitle = fig._suptitle
-        if suptitle and suptitle.get_in_layout():
-            rel_suptitle_height = fig.transFigure.inverted().transform_bbox(
-                suptitle.get_window_extent(renderer)).height
-            margin_top += rel_suptitle_height + pad_inches / fig_height_inch
+        suplabel = fig._suptitle
```
Comment (QuLogic) [inline lib/matplotlib/tight_layout.py:123]: Not sure if this one should be renamed.

File: lib/matplotlib/_constrained_layout.py [hunk 5]
```diff
@@ -276,21 +277,35 @@ def _make_layout_margins(fig, renderer, *, w_pad=0, h_pad=0,
...
-        w_pad, one = (fig.transSubfigure -
-                      parenttrans).transform((w_pad, 1))
-        h_pad = one - h_pad
-        bbox = invTransFig(fig._suptitle.get_tightbbox(renderer))
         p = fig._suptitle.get_position()
-        fig._suptitle.set_position((p[0], 1-h_pad))
-        fig._layoutgrid.edit_margin_min('top', bbox.height + 2 * h_pad)
+        fig._suptitle.set_position((p[0], 1 - h_pad_local))
+        bbox = invTransFig(fig._suptitle.get_tightbbox(renderer))  # fig coords
+        fig._layoutgrid.edit_margin_min('top', bbox.height + 2.0 * h_pad)
```
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:296]: Small inconsistency.
```suggestion
        fig._layoutgrid.edit_margin_min('top', bbox.height + 2 * h_pad)
```

File: doc/users/next_whats_new/suplabels.rst [hunk 6]
```diff
@@ -0,0 +1,18 @@
+suplabelx and suplabely
```
Comment (QuLogic) [inline doc/users/next_whats_new/suplabels.rst:1]:
```suggestion
supxlabel and supylabel
```

File: lib/matplotlib/figure.py [hunk 7]
```diff
@@ -436,19 +439,48 @@ def suptitle(self, t, **kwargs):
...
+
+    def suptitle(self, t, **kwargs):
+        # docstring from _suplabels...
+        info = {'name': '_suptitle', 'x0': 0.5, 'y0': 0.98,
+                'ha': 'center', 'va': 'top', 'rotation': 0}
+        return self._suplabels(t, info, **kwargs)
+
+    if suptitle.__doc__:
+        suptitle.__doc__ = _suplabels.__doc__.format(
+            x0=0.5, y0=0.98, name='suptitle', ha='center', va='top')
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/matplotlib/figure.py:463]: At the very least, you need to do `if _suplabels.__doc__;`, because `subtitle.__doc__ is None` since nothing is set on it yet.

File: examples/subplots_axes_and_figures/figure_title.py [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
+=============================================
+Figure labels: suptitle, supxlabel, supylabel
+=============================================
 
-Each subplot can have its own title (`.Axes.set_title`). Additionally,
-`.Figure.suptitle` adds a centered title at the top of the figure.
+Often it is desirable to have a title for a figure, while each subplot gets
+its own title using `.FigureBase.suptitle`.
+
+We can also add figure-level supxlabel and supylabel
```
[the remark this answers is not in the document]
Comment (jklymak) [reply examples/subplots_axes_and_figures/figure_title.py:9]: Lets keep it so people may be more likely to read below...

File: lib/matplotlib/_constrained_layout.py (hunk 5 above)
Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:296]: I went the other way.  I know that non-floats are fashionable in python, but I've *really* been stumped a few times when an array went all integers, so I strongly prefer being explicit.

File: lib/matplotlib/figure.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/figure.py:463]: Sure, did the `@docstring` I didn't know about that..;

File: lib/matplotlib/figure.py [hunk 9]
```diff
@@ -436,19 +439,48 @@ def suptitle(self, t, **kwargs):
...
+        return self._suplabels(t, info, **kwargs)
+
+    if supxlabel.__doc__:
+        supxlabel.__doc__ = _suplabels.__doc__.format(
+            x0=0.5, y0=0.01, name='supxlabel', ha='center', va='bottom')
+
+    def supylabel(self, t, **kwargs):
+        # docstring from _suplabels...
+        info = {'name': '_supylabel', 'x0': 0.02, 'y0': 0.5,
+                'ha': 'left', 'va': 'center', 'rotation': 90}
```
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/figure.py:478]: Sure, lets try it...

File: lib/matplotlib/figure.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/figure.py]: Oh, hum.   So `@docstring.copy` is something we hand-roll, and is only used for quiver?  Do we want to make it standard?  `self._suplabels` doesn't exist, so it needs to be used differently than above...

File: doc/users/next_whats_new/suplabels.rst (hunk 6 above)
Comment (jklymak) [reply doc/users/next_whats_new/suplabels.rst:1]: oops!

File: lib/matplotlib/_constrained_layout.py (hunk 5 above)
Comment (QuLogic) [reply lib/matplotlib/_constrained_layout.py:296]: There's still one on line 308, then

File: lib/matplotlib/tests/test_figure.py [hunk 10]
```diff
@@ -838,7 +838,7 @@ def test_reused_gridspec():
                   savefig_kwarg={'facecolor': 'teal'},
                   remove_text=False)
 def test_subfigure():
-    np.random.seed(19680808)
+    np.random.seed(19680801)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:841]: Were the images being updated already? Maybe no need to change this one otherwise.

File: lib/matplotlib/tests/test_figure.py [hunk 11]
```diff
@@ -862,7 +862,7 @@ def test_subfigure():
                   remove_text=False)
 def test_subfigure_ss():
     # test assigning the subfigure via subplotspec
-    np.random.seed(19680808)
+    np.random.seed(19680801)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:865]: Ditto.

File: lib/matplotlib/tests/test_figure.py (hunk 10 above)
Comment (jklymak) [reply lib/matplotlib/tests/test_figure.py:841]: Ooops, good point.  Like they aren't in a release yet, but yes they are in the repo...

File: examples/subplots_axes_and_figures/figure_title.py [hunk 12]
```diff
@@ -24,4 +28,36 @@
 
 fig.suptitle('Different types of oscillations', fontsize=16)
 
+##############################################################################
+# Note it is also sometimes useful to have a global x- or y-label, which
+# can be achieved with the `.FigureBase.supxlabel` and `.FigureBase.supylabel`
+# methods.
+
+fig, axs = plt.subplots(3, 5, figsize=(8, 5), constrained_layout=True,
+                        sharey=True, sharex=True)
```
[the remark this answers is not in the document]
Comment (jklymak) [reply examples/subplots_axes_and_figures/figure_title.py:38]: ... sure...

File: lib/matplotlib/tests/test_figure.py (hunk 10 above)
Comment (jklymak) [reply lib/matplotlib/tests/test_figure.py:841]: OK< so they did change.  Even if I kept the old (incorrect) seed, there was a slight horizontal shift, I assume due to the slot in place for the supylabel.

Comment (QuLogic) [reply lib/matplotlib/tests/test_figure.py:841]: OK, seems fine to change then.

**Review by QuLogic — APPROVED** [review]

File: examples/subplots_axes_and_figures/figure_title.py [hunk 13]
```diff
@@ -1,12 +1,17 @@
...
-Figure title
-============
+=============================================
+Figure labels: suptitle, supxlabel, supylabel
+=============================================
 
-Each subplot can have its own title (`.Axes.set_title`). Additionally,
-`.Figure.suptitle` adds a centered title at the top of the figure.
+Often it is desirable to have a title for a figure, while each subplot gets
+its own title using `.FigureBase.suptitle`.
```
Comment (timhoffm) [inline examples/subplots_axes_and_figures/figure_title.py:7]: This is a somewhat misleading wording. Gives the impression that `suptitle` would create the indivicual subplot titles.

File: examples/subplots_axes_and_figures/figure_title.py [hunk 14]
```diff
@@ -24,4 +29,36 @@
 
 fig.suptitle('Different types of oscillations', fontsize=16)
 
+##############################################################################
+# Note it is also sometimes useful to have a global x- or y-label, which
+# can be achieved with the `.FigureBase.supxlabel` and `.FigureBase.supylabel`
+# methods.
```
Comment (timhoffm) [inline examples/subplots_axes_and_figures/figure_title.py:35]: Same information, less wordy:
```suggestion
# A global x- or y-label can be set using the `.FigureBase.supxlabel` and
# `.FigureBase.supylabel` methods.
```

File: lib/matplotlib/_constrained_layout.py [hunk 15]
```diff
@@ -276,21 +277,35 @@ def _make_layout_margins(fig, renderer, *, w_pad=0, h_pad=0,
 def _make_margin_suptitles(fig, renderer, *, w_pad=0, h_pad=0):
     # Figure out how large the suptitle is and make the
     # top level figure margin larger.
+
+    invTransFig = fig.transFigure.inverted().transform_bbox
```
Comment (timhoffm) [inline lib/matplotlib/_constrained_layout.py:281]: Since you're essentially rewriting the whole function, would you care to make the name PEP-8 compliant?

Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:281]: The function name?  Or `invTransFig`?

File: examples/subplots_axes_and_figures/figure_title.py (hunk 13 above)
Comment (jklymak) [reply examples/subplots_axes_and_figures/figure_title.py:7]: Goops, yes, I'm confused what I thought I was saying here at all!

File: lib/matplotlib/_constrained_layout.py (hunk 15 above)
Comment (timhoffm) [reply lib/matplotlib/_constrained_layout.py]: `invTransFig`

Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py]: Done..

**Comment by dopplershift** [comment]
Were we relying on Travis to run any unique test configurations? I'm wondering why test lines execute has dropped so much...

**Comment by QuLogic** [comment]
Is it not posting? It seems to have run...

Not everything is installed in Azure Pipelines, so I don't think it covers the same things.

**Comment by dopplershift** [comment]
Travis CI doesn't show up in the current list of checks on this PR. It looks like no PR builds have run in 3 days. I'm assuming we're running into the same thing that SciPy and NumPy have run into due to Travis changing their pricing model:

image

**Comment by QuLogic** [comment]
Oh, I did not notice the date. I guess we'll put this back on the meeting agenda.

**Review by tacaswell — APPROVED** [review]
This is very nice looking!

I also really like the test images, they make it very clear what is going on.

**Comment by tacaswell** [comment]
Anyone can merge on CI green.

**Comment by QuLogic** [comment]
Code coverage is normal again, since this ran with GitHub Actions after @tacaswell's commit.

**Comment by jklymak** [comment]
Thanks for everyone's reviews, and for @QuLogic for fixing codecov!
